File: code/platform/src/main/java/org/fbme/ide/platform/testing/PlatformTestRunner.java
package org.fbme.ide.platform.testing;

import jetbrains.mps.module.ReloadableModule;
import jetbrains.mps.smodel.MPSModuleRepository;
import jetbrains.mps.smodel.ModelAccessHelper;
import jetbrains.mps.tool.environment.Environment;
import jetbrains.mps.tool.environment.EnvironmentConfig;
import jetbrains.mps.tool.environment.MpsEnvironment;
import org.junit.runner.notification.RunNotifier;
import org.junit.runners.BlockJUnit4ClassRunner;
import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.InitializationError;
import org.junit.runners.model.Statement;

import java.io.File;
import java.lang.reflect.Field;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Path;
import java.util.Objects;
import java.util.stream.StreamSupport;

public class PlatformTestRunner extends BlockJUnit4ClassRunner {

    private static volatile RunnerEnvironment ourRunnerEnvironment;

    public PlatformTestRunner(Class<?> clazz) throws InitializationError {
        super(loadFromCustomClassloader(clazz));
    }

    private static Class<?> loadFromCustomClassloader(Class<?> clazz) throws InitializationError {
        try {
            initializeRunnerEnvironment();
            return Class.forName(clazz.getName(), true, ourRunnerEnvironment.classLoader);
        } catch (ClassNotFoundException | MalformedURLException | NoSuchFieldException e) {
            throw new InitializationError(e);
        }
    }

    private static void initializeRunnerEnvironment() throws MalformedURLException, ClassNotFoundException, NoSuchFieldException {
        if (ourRunnerEnvironment == null) {
            synchronized (PlatformTestRunner.class) {
                if (ourRunnerEnvironment == null) {
                    String[] additionalLibs = System.getProperty("org.fbme.testing.additionalLibs", "").split(";");
                    String moduleName = System.getProperty("org.fbme.testing.module");

                    EnvironmentConfig config = EnvironmentConfig.defaultConfig()
                            .addLib(Path.of("../platform/build/artifacts/fbme_platform/fbme.platform/languages").toAbsolutePath().normalize().toString())
                            .addLib(Path.of("../library/build/artifacts/fbme_library/fbme.library/languages").toAbsolutePath().normalize().toString())
                            .setCreatePluginClassLoaders(false)
                            .withTestModeOn();
                    for (String lib: additionalLibs) {
                        config.addLib(lib);
                    }
                    MpsEnvironment environment = new MpsEnvironment(config);
                    environment.init();

                    MPSModuleRepository repository = MPSModuleRepository.getInstance();
                    ReloadableModule parentModule = new ModelAccessHelper(repository.getModelAccess()).runReadAction(() -> {
                        ReloadableModule platformModule = (ReloadableModule) StreamSupport.stream(repository.getModules().spliterator(), false)
                                .filter(it -> Objects.equals(it.getModuleName(), moduleName))
                                .findFirst().orElseThrow();
                        try {
                            platformModule.getClass("org.fbme.ide.iec61499.repository.MpsBridgeImpl").getMethod("install").invoke(null);
                        } catch (Exception e) {
                            throw new RuntimeException("Bridge not installed", e);
                        }
                        return platformModule;
                    });

                    PlatformTestClassloader classloader = new PlatformTestClassloader(parentModule);

                    Class<?> platformTestClass =
                            Class.forName("org.fbme.ide.platform.testing.PlatformTestBase", true, classloader);
                    ourRunnerEnvironment = new RunnerEnvironment(
                            environment,
                            classloader,
                            platformTestClass,
                            platformTestClass.getField("myEnvironment")
                    );
                }
            }
        }
    }

    @Override
    protected Statement withBefores(FrameworkMethod method, Object target, Statement statement) {
        Statement original = super.withBefores(method, target, statement);
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                if (ourRunnerEnvironment.platformTestClass.isInstance(target)) {
                    ourRunnerEnvironment.platformTestEnvironmentField.set(target, ourRunnerEnvironment.environment);
                }
                original.evaluate();
            }
        };
    }

    // Runs junit tests in a separate thread using the custom class loader
    @Override
    public void run(final RunNotifier notifier) {
        Runnable runnable = () -> {
            super.run(notifier);
        };
        Thread thread = new Thread(runnable);
        thread.setContextClassLoader(ourRunnerEnvironment.classLoader);
        thread.start();
        try {
            thread.join();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    private static class RunnerEnvironment {
        private final Environment environment;
        private final PlatformTestClassloader classLoader;
        private final Class<?> platformTestClass;
        private final Field platformTestEnvironmentField;

        public RunnerEnvironment(Environment environment, PlatformTestClassloader classLoader, Class<?> platformTestClass, Field platformTestEnvironmentField) {
            this.environment = environment;
            this.classLoader = classLoader;
            this.platformTestClass = platformTestClass;
            this.platformTestEnvironmentField = platformTestEnvironmentField;
        }
    }

    private static class PlatformTestClassloader extends URLClassLoader {

        private final ReloadableModule myParentModule;

        public PlatformTestClassloader(ReloadableModule parentModule) throws MalformedURLException {
            super(
                    new URL[]{
                            new File("build/classes/java/test").toURI().toURL(),
                            new File("build/resources/test").toURI().toURL()
                    },
                    null
            );
            myParentModule = parentModule;
        }

        @Override
        public Class<?> loadClass(String name) throws ClassNotFoundException {
            try {
                return super.loadClass(name);
            } catch (ClassNotFoundException ignored) {
            }
            return myParentModule.getClass(name);
        }
    }
}

File: code/platform/src/main/kotlin/org/fbme/ide/platform/testing/PlatformTestRunner.kt
package org.fbme.ide.platform.testing

import jetbrains.mps.module.ReloadableModule
import jetbrains.mps.smodel.MPSModuleRepository
import jetbrains.mps.smodel.ModelAccessHelper
import jetbrains.mps.tool.environment.Environment
import jetbrains.mps.tool.environment.EnvironmentConfig
import jetbrains.mps.tool.environment.IdeaEnvironment
import jetbrains.mps.util.PathManager
import org.junit.runner.notification.RunNotifier
import org.junit.runners.BlockJUnit4ClassRunner
import org.junit.runners.model.FrameworkMethod
import org.junit.runners.model.InitializationError
import org.junit.runners.model.Statement
import java.io.File
import java.lang.reflect.Field
import java.net.MalformedURLException
import java.net.URLClassLoader
import java.nio.file.Path
import java.util.concurrent.ConcurrentHashMap

class PlatformTestRunner(clazz: Class<*>) : BlockJUnit4ClassRunner(loadFromCustomClassloader(clazz)) {

    private val runnerEnvironment = runnerEnvironments.getValue(loadFromModuleName(clazz))

    override fun withBefores(method: FrameworkMethod, target: Any, statement: Statement): Statement {
        val original = super.withBefores(method, target, statement)
        return object : Statement() {
            @Throws(Throwable::class)
            override fun evaluate() {
                original.evaluate()
            }
        }
    }

    // Runs junit tests in a separate thread using the custom class loader
    override fun run(notifier: RunNotifier) {
        val runnable = Runnable { super.run(notifier) }
        val thread = Thread(runnable)
        thread.contextClassLoader = runnerEnvironment.classLoader
        thread.start()
        try {
            thread.join()
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }

    class RunnerEnvironment(val classLoader: PlatformTestClassloader)

    class PlatformTestClassloader(private val myParentModule: ReloadableModule) : URLClassLoader(
        arrayOf(
            File("build/classes/java/test").toURI().toURL(),
            File("build/classes/kotlin/test").toURI().toURL(),
            File("build/resources/test").toURI().toURL()
        ),
        null
    ) {
        @Throws(ClassNotFoundException::class)
        override fun loadClass(name: String): Class<*> {
            try {
                return super.loadClass(name)
            } catch (ignored: LinkageError) {
            } catch (ignored: ClassNotFoundException) {
            }
            return myParentModule.getClass(name)
        }
    }

    companion object {
        private val runnerEnvironments = ConcurrentHashMap<String, RunnerEnvironment>()

        @Throws(InitializationError::class)
        private fun loadFromCustomClassloader(clazz: Class<*>): Class<*> {
            return try {
                val moduleName = loadFromModuleName(clazz)
                initializeRunnerEnvironment(moduleName)
                Class.forName(clazz.name, true, runnerEnvironments.getValue(moduleName).classLoader)
            } catch (e: ClassNotFoundException) {
                throw InitializationError(e)
            } catch (e: MalformedURLException) {
                throw InitializationError(e)
            } catch (e: NoSuchFieldException) {
                throw InitializationError(e)
            }
        }

        private fun loadFromModuleName(clazz: Class<*>): String {
            val loadFrom = clazz.getAnnotation(LoadFrom::class.java)
            return loadFrom?.module ?: "org.fbme.ide.platform"
        }

        @Throws(MalformedURLException::class, ClassNotFoundException::class, NoSuchFieldException::class)
        fun initializeRunnerEnvironment(moduleName: String) {
            runnerEnvironments.getOrPut(moduleName) {
                val config = EnvironmentConfig.defaultConfig()
                    .withBuildPlugin()
                    .addDistributedPlugin("mps-execution-api", "jetbrains.mps.execution.api")
                    .addDistributedPlugin("mps-execution-configurations", "jetbrains.mps.execution.configurations")
                    .addFbmePlugin("fbme.library")
                    .addFbmePlugin("fbme.language")
                    .addFbmePlugin("fbme.platform")
                    .addFbmePlugin("fbme.richediting")
                    .addFbmePlugin("fbme.scenes")
                    .addFbmePlugin("fbme.formalfb")
                    .addFbmePlugin("fbme.integration.nxt")
                val environment = IdeaEnvironment(config)
                environment.init()
                val repository = MPSModuleRepository.getInstance()
                val parentModule = ModelAccessHelper(repository.modelAccess).runReadAction<ReloadableModule> {
                    val platformModule = repository.modules.first { it.moduleName == moduleName } as ReloadableModule
                    try {
                        platformModule
                            .getClass("org.fbme.ide.iec61499.repository.MpsBridgeImpl")
                            .getMethod("install")
                            .invoke(null)
                    } catch (e: Exception) {
                        throw RuntimeException("Bridge not installed", e)
                    }
                    platformModule
                }
                val classloader = PlatformTestClassloader(parentModule)
                val platformTestClass =
                    Class.forName("org.fbme.ide.platform.testing.PlatformTestBase", true, classloader)
                val field = platformTestClass.getDeclaredField("environment")
                field.isAccessible = true
                field.set(null, environment)
                RunnerEnvironment(classloader)
            }
        }

        private fun libPath(relative: String) = Path.of(relative).toAbsolutePath().normalize().toString()

        private fun EnvironmentConfig.addDistributedPlugin(folder: String, id: String): EnvironmentConfig {
            val preinstalledPluginFolder = File(PathManager.getPreInstalledPluginsPath())
            return addPlugin(File(preinstalledPluginFolder, folder).absolutePath, id)
        }

        private fun EnvironmentConfig.addFbmePlugin(id: String): EnvironmentConfig {
            return addPlugin(libPath("../../build/dist-plugins/$id"), id)
        }

        init {
            System.setProperty("ide.widget.toolbar", "false")
            System.setProperty("ide.cancellation.propagate", "false")
        }
    }
}
